File: WebGL/NodeJS/public/src/Primitives/Cone.js

class Cone extends Primitive3D
{
	constructor()
	{
		super(Primitive3D.CONE);
		
		this._r = 0;
		this._h = 0;
	}
	
	async initWithPosAndSize(x, y, z, r, h)
	{
		this.m[12] = x;
		this.m[13] = y;
		this.m[14] = z;
		
		this._r = r;
		this._h = h;
		
		this.initCommon();
	}
	
	async initWithMatAndSize(mat, r, h)
	{
		for(var i=0; i<16; i++)
		{
			this.m[i] = mat[i];
		}

		this._r = r;
		this._h = h;

		this.initCommon();
	}

	async initWithCone(cone)
	{
		for(var i=0; i<16; i++)
		{
			this.m[i] = cone.m[i];
		}

		this._r = cone.getRadius();
		this._h = cone.getHeight();
		this._id = cone.getID();

		this.initCommon();
	}
	
	async initCommon()
	{
		this._vertexBufferID = 0;
		this._colorBufferID = 0;
		this._vertexCount = 0;
		this._randomColor = new RandomColor();

		this._shaderProgram = new ShaderProgram();
		await this._shaderProgram.init("shaders/Primitive3D/Primitive3D.vs", "shaders/Primitive3D/Primitive3D.fs");

		this.generateBufferID();
	}
	
	getPos()
	{
		return new CVector3(this.m[12], this.m[13], this.m[14]);
	}

	getRadius()	{ return this._r; }
	getHeight()	{ return this._h; }

	setRadius(r)
	{ 
		if(r > 0)
			this._r = r; 
	}
	
	setHeight(h)
	{ 
		if(h > 0)
			this._h = h;
	}

	volume()
	{
		return (1.0/3.0) * PI_VAL * this._r * this._r * this._h;
	}
	

	generateBufferID()
	{
		var buffer = new GLBuffer(true, false, false);

		var radius = 0.5;
		var halfLength = 0.5;
		var piVal = 3.14159265;

		buffer.glBegin(gl.TRIANGLES);

		//if(_useRandomColors)
		//	_randomColor.Reset();

		for(var i=20; i<=360; i+=20)
		{
			var theta = i*piVal/180.0;
			var nextTheta = (i+20)*piVal/180.0;

			if(this._useRandomColors)
				buffer.glColoruia(this._randomColor.nextColor(), this._randomColorAlpha);

			buffer.glVertex3f(0, halfLength, 0);

			buffer.glColor4ub(80, 80, 80, 255);
			buffer.glVertex3f(radius*Math.cos(nextTheta),	-halfLength,	radius*Math.sin(nextTheta));
			buffer.glVertex3f(radius*Math.cos(theta),		-halfLength,	radius*Math.sin(theta));

			buffer.glVertex3f(0, -halfLength, 0);

			if(this._useRandomColors)
				buffer.glColoruia(this._randomColor.nextColor(), this._randomColorAlpha);

			buffer.glVertex3f(radius*Math.cos(theta),		-halfLength,	radius*Math.sin(theta));
			buffer.glVertex3f(radius*Math.cos(nextTheta),	-halfLength,	radius*Math.sin(nextTheta));
		}

		buffer.glEnd();

		this._vertexBufferID = buffer.getVertexBufferID();
		this._colorBufferID = buffer.getColorBufferID();
		this._vertexCount = buffer.getVertexCount();
	}
	

	draw()
	{
		if(!this._visible)
			return;

		this._scaleMat[0] = this._r * 2;
		this._scaleMat[5] = this._h;
		this._scaleMat[10] = this._r * 2;

		this._shaderProgram.begin();
		
		var projMatLoc = gl.getUniformLocation(this._shaderProgram.programID, "projMat");
		var modelMatLoc = gl.getUniformLocation(this._shaderProgram.programID, "modelMat");
		var oriMatLoc = gl.getUniformLocation(this._shaderProgram.programID, "oriMat");
		var scaleMatLoc = gl.getUniformLocation(this._shaderProgram.programID, "scaleMat");
				
		gl.uniformMatrix4fv(projMatLoc, false, cam3D.projMat.m);
		gl.uniformMatrix4fv(modelMatLoc, false, cam3D.modelMat.m);
		gl.uniformMatrix4fv(oriMatLoc, false, this.m);
		gl.uniformMatrix4fv(scaleMatLoc, false, this._scaleMat);
	

		var colorID = gl.getAttribLocation(this._shaderProgram.programID, "color");
		gl.enableVertexAttribArray(colorID);
		gl.bindBuffer(gl.ARRAY_BUFFER, this._colorBufferID);
		gl.vertexAttribPointer( colorID, 4, gl.UNSIGNED_BYTE, gl.FALSE, 0, 0);

		var vertexID = gl.getAttribLocation(this._shaderProgram.programID, "vertex");
		gl.enableVertexAttribArray(vertexID);
		gl.bindBuffer(gl.ARRAY_BUFFER, this._vertexBufferID);
		gl.vertexAttribPointer(vertexID, 3, gl.FLOAT, gl.FALSE, 0, 0);

		gl.drawArrays(gl.TRIANGLES, 0, this._vertexCount);
		
		gl.bindBuffer(gl.ARRAY_BUFFER, null);
		gl.disableVertexAttribArray(vertexID);
		gl.disableVertexAttribArray(colorID);

		this._shaderProgram.end();
	}
	
}

File: GLSL/ObjLoader/source/Shapes3D/Cone.cpp
#include "Cone.h"
#include "math.h"
#include "GLUtil.h"
#include "TransformVertexBuf.h"
#include "Cylinder.h"
#include "ShadersManager.h"

Cone::Cone() : Shape(Shape::CONE)
{
	_r = 0;
	_h = 0;

	InitCommon();
}


Cone::Cone(float* mat, float r, float h) : Shape(Shape::CONE)
{
	memcpy(m, mat, 16*4);

	_r = r;
	_h = h;

	InitCommon();
}


Cone::Cone(float x, float y, float z, float r, float h) : Shape(Shape::CONE)
{
	m[12] = x;
	m[13] = y;
	m[14] = z;

	_r = r;
	_h = h;

	InitCommon();
}


Cone::Cone(Cone* cone)
{
	memcpy(m, cone->m, 16*4);

	_r = cone->GetRadius();
	_h = cone->GetHeight();
	_id = cone->GetID();

	InitCommon();
}


void Cone::InitCommon()
{
	_vertexBufferID = 0;
	_colorBufferID = 0;

	_vertexCount = 0;

	_shaderProgram = ShadersManager::GetInstance()->CreateShaderProgram("shaders/ColorArray/ColorArray.vs", "shaders/ColorArray/ColorArray.fs");

	GenerateBufferID();
}


CVector3 Cone::GetPos()
{
	return CVector3(m[12], m[13], m[14]);
}


float Cone::GetRadius()			{ return _r; }
float Cone::GetHeight()			{ return _h; }

void Cone::SetRadius(float r)	
{ 
	if(r > 0)
		_r = r; 
}
void Cone::SetHeight(float h)
{ 
	if(h > 0)
		_h = h;
}


float Cone::Volume()
{
	return (1.0f/3.0f) * PI_VAL * _r * _r * _h;
}


Cone Cone::CalcBoundingCone(float* vertexBuf, int arrSize)
{
	float* localVertexBuf = new float[arrSize];
	memcpy(localVertexBuf, vertexBuf, arrSize*sizeof(float));

	Cylinder cylinder = Cylinder::CalcBoundingCylinder(vertexBuf, arrSize);

	float cylMatInvert[16];
	GLMat::InvertMatrix(cylinder.GetGLMatrix(), cylMatInvert);

	cylMatInvert[13] += cylinder.GetHeight()/2.0f;

	TransformVertexBuf::MulBufWithMatrix(localVertexBuf, arrSize, cylMatInvert);

	float height = cylinder.GetHeight() * 1.05f;
	float radius = CalcRadius(localVertexBuf, arrSize, height);
	float volume = (1.0f/3.0f) * PI_VAL * radius * radius * height;

	float finalH = height;

	bool addX180Rot = false;

	for(int loop=0; loop<2; loop++)
	{
		if(loop == 1)
		{
			float cylinderH = cylinder.GetHeight();
			
			for(int i=3; i<arrSize; i+=3)
				localVertexBuf[i+1] = cylinderH - localVertexBuf[i+1];
		}

		for(float i=1.1f; i<2.0f; i+=0.1f)
		{
			float currH = height * i;

			float r = CalcRadius(localVertexBuf, arrSize, currH);
			float v = (1.0f/3.0f) * PI_VAL * r * r * currH;

			if(v < volume)
			{
				volume = v;
				radius = r;
				finalH = currH;

				if(loop == 1)
					addX180Rot = true;
			}
		}
	}

	delete[] localVertexBuf;

	float matInvert[16];
	GLMat::InvertMatrix(cylMatInvert, matInvert);

	GLMat mat;
	mat.glMultMatrixf(matInvert);

	if(addX180Rot)
	{
		mat.glTranslatef(0,cylinder.GetHeight(),0);
		mat.glRotatef(180, 1,0,0);
	}

	mat.glTranslatef(0,finalH/2,0);

	Cone cone(mat.m, radius, finalH);

	return cone;
}


float Cone::CalcRadius(float* vertexBuf, int arrSize, float height)
{
	float x = vertexBuf[0];
	float y = vertexBuf[1];
	float z = vertexBuf[2];

	float dy = abs(height - y);
	float xOnXZPlane = height * x / dy;
	float zOnXZPlane = height * z / dy;

	float rPow2 = xOnXZPlane*xOnXZPlane + zOnXZPlane*zOnXZPlane;

	for(int i=3; i<arrSize; i+=3)
	{
		x = vertexBuf[i+0];
		y = vertexBuf[i+1];
		z = vertexBuf[i+2];

		dy = height - y;
		xOnXZPlane = height * x / dy;
		zOnXZPlane = height * z / dy;

		float dist = xOnXZPlane*xOnXZPlane + zOnXZPlane*zOnXZPlane;

		if(rPow2 < dist)
			rPow2 = dist;
	}

	return sqrt(rPow2);
}


void Cone::Draw()
{
	if(!_visible)
		return;

	glPushMatrix();
	glMultMatrixf(m);

	glScalef(_r*2, _h, _r*2);

	_shaderProgram->Begin();

	GLuint colorID = glGetAttribLocation(_shaderProgram->ProgramID(), "color");
	glEnableVertexAttribArray(colorID);
	glBindBuffer(GL_ARRAY_BUFFER, _colorBufferID);
	glVertexAttribPointer( colorID, 4, GL_UNSIGNED_BYTE, GL_TRUE, 0, (void*)0);

	GLuint vertexID = glGetAttribLocation(_shaderProgram->ProgramID(), "vertex");
	glEnableVertexAttribArray(vertexID);
	glBindBuffer(GL_ARRAY_BUFFER, _vertexBufferID);
	glVertexAttribPointer(vertexID, 3, GL_FLOAT, GL_TRUE, 0, (void*)0);

	glDrawArrays(GL_TRIANGLES, 0, _vertexCount);
	
	glBindBuffer(GL_ARRAY_BUFFER, 0);
	glDisableVertexAttribArray(vertexID);
	glDisableVertexAttribArray(colorID);

	_shaderProgram->End();

	glPopMatrix();
}

void Cone::GenerateBufferID()
{
	GLBuffer* buffer = new GLBuffer(100, true, false, false);

	float radius = 0.5f;
	float halfLength = 0.5f;
	float piVal = 3.14159265f;

	buffer->glBegin(GL_TRIANGLES);

	if(_useRandomColors)
		_randomColor.Reset();

	for(int i=20; i<=360; i+=20)
	{
		float theta = i*piVal/180.0f;
		float nextTheta = (i+20)*piVal/180.0f;

		if(_useRandomColors)
			buffer->glColor(_randomColor.NextColor());

		buffer->glVertex3f(0, halfLength, 0);

		buffer->glColor4ub(80, 80, 80, 255);
		buffer->glVertex3f(radius*cos(nextTheta),	-halfLength,	radius*sin(nextTheta));
		buffer->glVertex3f(radius*cos(theta),	-halfLength,	radius*sin(theta));

		buffer->glVertex3f(0, -halfLength, 0);

		if(_useRandomColors)
			buffer->glColor(_randomColor.NextColor());

		buffer->glVertex3f(radius*cos(theta),	-halfLength,	radius*sin(theta));
		buffer->glVertex3f(radius*cos(nextTheta),	-halfLength,	radius*sin(nextTheta));
	}

	buffer->glEnd();

	_vertexBufferID = buffer->GetVertexBufferID();
	_colorBufferID = buffer->GetColorBufferID();

	_vertexCount = buffer->GetVertexCount();

	delete buffer;
}

Cone::~Cone()
{
	if (_shaderProgram != NULL)
	{
		string vertexShaderPath = _shaderProgram->GetVertexShaderFilePath();
		string fragementShaderPath = _shaderProgram->GetFragmentShaderFilePath();

		ShadersManager::GetInstance()->DeleteShaderProgram(_shaderProgram);
	}
}
